ShadowsocksX-NG/AppDelegate.swift

```diff
@@ -103,6 +103,15 @@ class AppDelegate: NSObject, NSApplicationDelegate, NSUserNotificationCenterDele
             "Kcptun.LocalPort": NSNumber(value: 8388),
             "Kcptun.Conn": NSNumber(value: 1),
             ])
+
+        // To keep the default exception values backward compatible, we set it
+        // to the previously hardcoded value if the key doesn't exist.
+        // Registering defaults won't work because the empty state is a valid
+        // value, and Cocoa Bindings won't distinguish between empty and string
+        // values.
+        if defaults.string(forKey: "ProxyExceptions") == nil {
+            defaults.set("127.0.0.1, localhost, 192.168.0.0/16, 10.0.0.0/8", forKey: "ProxyExceptions")
+        }
```

Comment:
Registering defaults 是工作的。你碰到问题是因为，你开发过程中老的版本曾经把这个值设置成空。
验证方法：
编辑文件~/Library/Preferences/com.qiuyuzhou.ShadowsocksX-NG.plist
将ProxyExceptions删除，再打开软件。

Reply:
Default value registering _mechanism_ itself works, but we can't utilize it to solve our problem, hence it _"won't work"_.

The preference dialog is implemented using Cocoa Bindings. It has an unexpected behavior: if you leave the text field empty, Cocoa Bindings will give `UserDefaults` a `nil` instead of `""`. So later, `defaults.string(forKey: "ProxyExceptions")` will give you the default value instead of an empty string, which is not the expected behavior.

You can reproduce this by setting the GFW List text field to empty, and reopening the preference dialog.

Reply:
Oh, I just saw your commit. Here's how to reproduce the problem:

If you set the __"Bypass proxy settings for these Hosts & Domains"__ text field to empty and close the preferences dialog:
* The corresponding system network settings will be `127.0.0.1, localhost, 192.168.0.0/16, 10.0.0.0/8` instead of empty.
* If you open the preferences dialog again, the text field of be `127.0.0.1, localhost, 192.168.0.0/16, 10.0.0.0/8` instead of empty.

Also, this makes the __Reset__ button useless.

Reply:
Oh, I see.

Reply:
After some thinking. I think that there is no need to support setting an empty string value.
If want implement it. A custom NSValueTransformer may be a better approach.